nuxeo-services/nuxeo-mail/src/main/java/org/nuxeo/mail/MailSessionBuilder.java

```diff
@@ -0,0 +1,227 @@
+
+        @Override
+        public String toString() {
+            ToStringBuilder builder = new ToStringBuilder(this);
+            Map<Object, Object> propertiesAsMap = properties.entrySet()
+                                                            .stream()
+                                                            .filter(s -> !s.getKey().toString().contains("password"))
+                                                            .collect(toMap(Map.Entry::getKey, Map.Entry::getValue));
+            builder.append("properties", propertiesAsMap);
+            return builder.toString();
+        }
+    }
+
+    @SuppressWarnings("rawtypes")
+    protected abstract static class AbstractFrom<F extends AbstractFrom> implements FromBuilder {
+
+        protected boolean debug;
+
+        /**
+         * Enables debug mode for built session.
+         *
+         * @implNote {@code mail.debug} in nuxeo.conf takes precedence if it's true.
+         * @return this builder
+         */
+        @SuppressWarnings("unchecked")
+        public F debug() {
+            this.debug = true;
+            return (F) this;
+        }
+
+        public Session build() {
+            log.info("Build a javax.mail.Session from builder: {}", this);
+            if (!debug) {
+                // check nuxeo.conf - same configuration key
+                log.trace("Mail log debug enabled by nuxeo.conf");
+                debug = Boolean.parseBoolean(Framework.getProperty(CONFIGURATION_MAIL_DEBUG, FALSE.toString()));
+            }
+            Session session = retrieveSession();
+            if (debug) {
+                session.setDebugOut(new PrintStream(new MailLogOutputStream()));
+                session.setDebug(true);
+            }
+            // set proper protocol for rfc822
+            String protocol = session.getProperty("mail.transport.protocol");
+            if (isNotEmpty(protocol)) {
+                session.setProtocolForAddress("rfc822", protocol);
+            }
+            return session;
+        }
+
+        protected abstract Session retrieveSession();
+
+        @Override
+        public Store buildAndConnect() {
+            try {
+                Session session = build();
+                Store store = session.getStore();
+                // handle backward compatibility for MailCoreHelper
+                // by default user/password are read from authenticator configured throw jndi Resource or Properties
+                String user = session.getProperty("user");
+                String password = session.getProperty("password");
+                store.connect(user, password); // accept nulls
+                return store;
+            } catch (MessagingException e) {
+                throw new NuxeoException("Unable to build/connect javax.mail.Store", e);
+            }
+        }
+    }
+
+    public interface FromBuilder {
+
+        @NotNull
+        Session build();
+
+        @NotNull
+        Store buildAndConnect();
```

Comment:
`connect` is not enough because to connect we want the build no ?

Reply:
Exactly 👍

It's just to let the caller know he's building something before connecting.